File: src/extensibility/python-extensibility/nexus_extensibility/_data_model_extensions.py
from datetime import timedelta

_quotients = [1000, 1000, 60, 1 ]
_post_fixes = ["us", "ms", "s", "min"]

def to_unit_string(sample_period: timedelta) -> str:
    """
    Converts period into a human readable number string with unit.

    Args:
        sample_period: The period to convert.
    """

    current_value = sample_period.total_seconds() * 1e6
    
    for i in range(len(_post_fixes)):

        quotient, remainder = divmod(current_value, _quotients[i])

        if remainder != 0:
            return f"{str(int(current_value))}_{_post_fixes[i]}"

        else:
            current_value = quotient

    return f"{str(int(current_value))}_{_post_fixes[-1]}"

File: src/extensibility/rust-extensibility/src/data_model/data_model_extensions.rs
use super::SamplePeriod;

static QUOTIENTS: [u128; 7] = [1000, 1000, 1000, 60, 60, 24, 1];
static POST_FIXES: [&str; 7] = ["ns", "us", "ms", "s", "min", "h", "d"];

/// Contains extension methods to make life easier working with the data model types.
pub struct DataModelExtensions;

impl DataModelExtensions {
    /// Converts the period into a human readable number string with unit.
    pub fn to_unit_string(sample_period: &SamplePeriod) -> String {
        let mut current_value = sample_period.as_ref().as_nanos();

        for i in 0..POST_FIXES.len() {
            let quotient = current_value / QUOTIENTS[i];
            let remainder = current_value % QUOTIENTS[i];

            if remainder != 0 {
                return format!("{}_{}", current_value, POST_FIXES[i]);
            } else {
                current_value = quotient;
            }
        }

        format!("{}_{}", current_value, POST_FIXES[POST_FIXES.len() - 1])
    }
}
